File: leetcode/_143_ReorderList.java
/**
 * @author Shahidul Islam
 * @date 1/25/2022
 **/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class _143_ReorderList {
    public void reorderList(ListNode head) {
        int n =0;
        ListNode cur = head;
        while (cur != null){
            cur = cur.next;
            n++;
        }

        cur = head;
        for (int i = 0; i < Math.ceil(n/2.0);i++){
            cur = cur.next;
        }
    }
}

File: leetcode/143. Reorder List.py
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        slow = head
        fast = slow.next if slow else None

        while fast and fast.next != None:
            slow = slow.next
            fast = fast.next.next

        sentinel = ListNode(-1, None)
        slow.next, slow = None, slow.next
        while slow:
            x = slow
            slow = slow.next
            x.next = sentinel.next
            sentinel.next = x
        slow = sentinel.next

        sentinel.next = None
        cur = sentinel
        while head:
            cur.next, head = head, head.next
            cur = cur.next
            cur.next = slow
            cur = cur.next
            if slow:
                slow = slow.next

        return sentinel.next
